feat(material-experimental/mdc-list): add test harness

devversion opened the pull request:
* Fixes a discovered issue with icon/avatar alignment in the list-option (that would otherwise block the harness).
* Adds a test harness for MDC-based list

**Note for reviewer**: I'd recommend reviewing on a per-commit basis.

Resolves COMP-337

devversion commented:
For some reason, the View Engine test fails. I need to look more into it.

mmalerba commented on src/material-experimental/mdc-list/list-option.ts:
```diff
@@ -233,5 +236,68 @@ export class MatListOption extends MatListItemBase implements OnInit, OnDestroy
+   * to another container. See: https://stackblitz.com/edit/angular-ivy-7veqgj
+   */
+  // * TODO: Consider removing if View Engine is no longer supported.
+  private _renderCheckboxAndIcons() {
```
I'm sad about the amount of additional code needed to work around this silly ViewEngine issue, but unfortunately I can't really think of a better way

devversion replied:
I spent more time looking into this (given the new changes for the icons/avatars) and was able to solve the issue in an Angular idiomatic way. Fortunately this block of code is no longer needed!

Initially I was under the impression it is an issue with the embedded view references (and their removal in View Engine), but it looks like View Engine just breaks due to the conditional use of multiple view refs in the same template outlet. The fix now is to just use two outlets (which make sense anyway now due to the new changes for icons/avatars).

StackBlitz showing the issue: https://stackblitz.com/edit/view-engine-container-template-outlet-issue
StackBlitz with the fix/workaround: https://stackblitz.com/edit/view-engine-workaround-template-outlet

mmalerba commented on src/material-experimental/mdc-list/testing/list-harness-base.ts:
```diff
@@ -0,0 +1,153 @@
+    let currentSection: ListSection<C> = {items: []};
+    const itemsAndSubheaders =
+        await this.getItemsWithSubheadersAndDividers({item: filters, divider: false});
+    for (const itemOrSubheader of itemsAndSubheaders) {
```
should this `for` be refactored into a `Promise.all`?

devversion replied:
Makes sense. I've updated the non-MDC list harness to also perform the header retrieval concurrently.

crisbeto commented on src/material-experimental/mdc-list/list-option.html:
```diff
@@ -29,24 +30,29 @@
 </span>
+<!-- Conditionally renders icons/avatars after the list item text. -->
+<ng-template [ngIf]="_shouldShowIconsAndAvatarsAt('after')">
+  <ng-template [ngTemplateOutlet]="icons"></ng-template>
```
I think you can drop the `ngIf` and the `ng-template` here like this:

```
<ng-container [ngTemplateOutlet]="_shouldShowIconsAndAvatarsAt('after') ? icons : null"></ng-container>
```

Same goes for the other outlets.

devversion replied:
Looks like this doesn't work and when the template outlet conditional changes, one outlet will not render anything. It might be related to timing of the embedded view that is shared between two outlets then. With a separate template it might avoid this problematic timing due to `ngIf` first being resolved and then the outlet being rendered afterwards.

crisbeto commented on src/material-experimental/mdc-list/list-option.ts:
```diff
@@ -185,24 +191,30 @@ export class MatListOption extends MatListItemBase implements OnInit, OnDestroy
   /** Whether the list-option has icons or avatars. */
   _hasIconOrAvatar() {
-    return this._avatars.length || this._icons.length;
+    return this._avatars.length !== 0 || this._icons.length !== 0;
```
Not something that was added in this PR, but I think that this may be prone to "changed after checked" errors since the `_avatars` and `_icons` are based on a content query.

devversion replied:
Yeah, seems out of scope for this PR. I haven't seen any issues with it yet, but should we look into it or keep as is?

crisbeto commented on src/material-experimental/mdc-list/list-styler-elements.ts:
```diff
@@ -0,0 +1,78 @@
+@Directive()
+export abstract class _MatListItemGraphicAlignmentBase {
+  @HostBinding('class.mdc-list-item__graphic')
+  get alignAtStart(): boolean {
```
Instead of having the two other directies extending this class, couldn't you change the selector so it matches the avatar and icon? That way you can get rid of the `extends`, the getters, and place the binding in `host`.

devversion replied:
This will require us to duplicate the directive selectors. I think both ways have trade-offs here, but I went with what you proposed since I'd like to avoid the getters.

crisbeto commented on src/material-experimental/mdc-list/testing/list-harness-base.ts:
```diff
@@ -0,0 +1,157 @@
+ * @template F The filter type used filter list item harness of type `C`.
+ * @docs-private
+ */
+export class MatListHarnessBase<
```
Should this be abstract?

devversion replied:
Yeah, sounds reasonable. This logic has been taken from the existing list harness, so I'll also update it there.

crisbeto commented on src/material-experimental/mdc-list/testing/selection-list-harness.ts:
```diff
@@ -0,0 +1,145 @@
+      return this.getItems();
+    }
+    return ([] as MatListOptionHarness[])
+        .concat(...await Promise.all(
```
Apart from casting the type, I don't see why this needs the `concat`. Isn't it the same as returning the `Promise.all` directly?

devversion replied:
This has also been taken from the existing list harness. I think the reason for it is to flatten the results from `locatorForAll`. I'm switching it to a reduce in both implementations for more readability and best-practices.

crisbeto commented on src/material-experimental/mdc-list/testing/selection-list-harness.ts:
```diff
@@ -0,0 +1,145 @@
+  private _beforeCheckbox = this.locatorForOptional('.mdc-list-item__graphic .mdc-checkbox');
+
+  /** Gets the position of the checkbox relative to the list option content. */
+  async getCheckboxPosition(): Promise<'before'|'after'> {
```
Should we have a type for `'before' | 'after'`? It seems to be repeated in a lot of places and it'll be a pain if we decide to add another one.

devversion replied:
I agree. Created a type for both implementations of the selection list. The non-MDC list harness has the same code.

devversion commented on src/material-experimental/mdc-list/testing/list-harness-base.ts:
```diff
@@ -0,0 +1,157 @@
+    T extends(ComponentHarnessConstructor<C>& {with: (options?: F) => HarnessPredicate<C>}),
+             C extends ComponentHarness, F extends BaseListItemHarnessFilters> extends
+    ComponentHarness {
+  protected _itemHarness: T;
```
This matches what we do in the existing list harness and I think passing it through the constructor will mean that we need to pass through parameters from the `ComponentHarness` (e.g. the locator factory or more in the future).

devversion commented:
@crisbeto Addressed most of your feedback. A few of the comments were about logic that has been just taken from the existing list harness. I addressed those that seemed reasonable and also applied the changes to the existing list harness (see separate commits)

mmalerba reviewed (approved):
LGTM

wagnermaciel commented:
@devversion Please resolve this merge conflict

devversion commented:
@wagnermaciel Done. Had to sync up with recent changes to the non-MDC list harnesses.
